### PR TITLE
Heap-allocate RpcEvents.

### PR DESCRIPTION
**Pull request by cschuet**
Replace Rpc's RpcEvent members with heap-allocated RpcEvents.

[RFC=0002](https://github.com/googlecartographer/rfcs/blob/master/text/0002-cloud-based-mapping-1.md)

### CODE REVIEW

File: cartographer_grpc/framework/rpc.h [hunk 1]
```diff
@@ -74,6 +71,7 @@ class Rpc {
       ::grpc::internal::RpcMethod::RpcType rpc_type);
   void SendFinish(std::unique_ptr<::google::protobuf::Message> message,
                   ::grpc::Status status);
+  bool* GetEventState(Event event);
```
Comment (gaschler) [inline cartographer_grpc/framework/rpc.h]: name consistency with setter: GetRpcEventState

File: cartographer_grpc/framework/rpc.h [hunk 2]
```diff
@@ -89,11 +87,16 @@ class Rpc {
...
   ::grpc::ServerContext server_context_;
 
-  RpcEvent new_connection_event_;
-  RpcEvent read_event_;
-  RpcEvent write_event_;
-  RpcEvent finish_event_;
-  RpcEvent done_event_;
+  // These state variables indicate whether the corresponding event is currently
+  // pending completion, e.g. 'read_event_pending_ = true' means that a read has
+  // been requested but hasn't completed yet. While
```
Comment (gaschler) [inline cartographer_grpc/framework/rpc.h]: Drop "While", drop comma.

File: cartographer_grpc/framework/service.cc [hunk 3]
```diff
@@ -131,10 +131,7 @@ void Service::HandleFinish(Rpc* rpc, bool ok) {
 void Service::HandleDone(Rpc* rpc, bool ok) { RemoveIfNotPending(rpc); }
 
 void Service::RemoveIfNotPending(Rpc* rpc) {
-  if (!rpc->GetRpcEvent(Rpc::Event::DONE)->pending &&
-      !rpc->GetRpcEvent(Rpc::Event::READ)->pending &&
-      !rpc->GetRpcEvent(Rpc::Event::WRITE)->pending &&
-      !rpc->GetRpcEvent(Rpc::Event::FINISH)->pending) {
+  if (rpc->IsNoEventPending()) {
```
Comment (gaschler) [inline cartographer_grpc/framework/service.cc]: alternative if you like it: !IsAnyEventPending

File: cartographer_grpc/framework/rpc.cc [hunk 4]
```diff
@@ -49,11 +49,11 @@ Rpc::Rpc(int method_index,
       execution_context_(execution_context),
       rpc_handler_info_(rpc_handler_info),
       service_(service),
-      new_connection_event_{Event::NEW_CONNECTION, this, false},
-      read_event_{Event::READ, this, false},
-      write_event_{Event::WRITE, this, false},
-      finish_event_{Event::FINISH, this, false},
-      done_event_{Event::DONE, this, false},
+      new_connection_event_pending_(false),
```
Comment (gaschler) [inline cartographer_grpc/framework/rpc.cc]: maybe defining these defaults in the header is slightly more readable?

**Comment by gaschler** [comment]
Also, consider a more concise commit message.

File: cartographer_grpc/framework/rpc.cc (hunk 4 above)
Comment (cschuet) [reply cartographer_grpc/framework/rpc.cc]: Done.

File: cartographer_grpc/framework/rpc.h (hunk 1 above)
Comment (cschuet) [reply cartographer_grpc/framework/rpc.h]: Done.

File: cartographer_grpc/framework/rpc.h (hunk 2 above)
Comment (cschuet) [reply cartographer_grpc/framework/rpc.h]: Done.

File: cartographer_grpc/framework/service.cc (hunk 3 above)
Comment (cschuet) [reply cartographer_grpc/framework/service.cc]: DOne.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**